### PR TITLE
refactor: removes use of fs.exists and updates mocks

### LINKED ISSUE #171: fs.exists is deprecated at Node.js >12
#### My reason:

We uses `fs.exists` at least twice in our source code, this has been deprecated, we should upgrade it.

https://nodejs.org/api/fs.html#fs_fs_exists_path_callback

### PR DESCRIPTION
**Pull request by juanmanual**
**Type:** Refactor

The following has been addressed in the PR:

*  There is a related issue? #171
*  Unit tests for the changed files have been verified and mocks updated as needed in the PR

**Description:**
Since fs.exists is deprecated, this pull request removes it from product source.

### CODE REVIEW

**Comment by juanpicado** [comment]
Thanks @juanmanual 👍  I'll review this.

File: src/local-fs.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ import { unlockFile, readFile } from '@verdaccio/file-locking';
 import { Callback, Logger, Package, ILocalPackageManager, IUploadTarball } from '@verdaccio/types';
 import { getCode, getInternalError, getNotFound, VerdaccioError } from '@verdaccio/commons-api/lib';
 
-export const fileExist = 'EEXISTS';
+export const fileExist = 'EEXIST';
```
Comment (juanpicado) [inline src/local-fs.ts]: Why this change?
If we modify that breaks the project here https://github.com/verdaccio/verdaccio/blob/master/src/lib/constants.ts#L161 since I am expecting such thing. It is not ideal to have suck weak relationship, but, I wonder why this change.

File: src/local-fs.ts [hunk 2]
```diff
@@ -286,11 +287,12 @@ export default class LocalFS implements ILocalFSPackageManager {
   private _createFile(name: string, contents: any, callback: Function) {
     this.logger.trace({ name }, '[local-storage/_createFile] create a new file: @{name}');
 
-    fs.exists(name, exists => {
-      if (exists) {
-        this.logger.trace({ name }, '[local-storage/_createFile] file cannot be created, it already exists: @{name}');
-
-        return callback(fSError(fileExist));
+    fs.open(name, 'wx', err => {
```
Comment (juanpicado) [inline src/local-fs.ts]: https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_file_system_flags

`'wx' Like 'w' but fails if the path exists.`

Could you explain me why this change?

Comment (juanpicado) [reply src/local-fs.ts]: Oh I see, nevermind, there was `fs.exists` before.

File: src/local-fs.ts (hunk 1 above)
Comment (juanpicado) [reply src/local-fs.ts]: Ok, here some context. The keyword `EEXISTS` a legacy from sinopia, it seems at Node.js `0.10` the codes were not documented and the first reference I've found by far is at Node.js 4.x https://nodejs.org/docs/latest-v4.x/api/errors.html , but this error handler is unrelated with the native one `EEXIST (File exists): An existing file was the target of an operation that required that the target not exist.`

At Verdaccio still we rely on `EEXISTS` so, this is super misleading, totally agree, but, we will change this in a future PR to avoid confusions

File: src/local-database.ts [hunk 3]
```diff
@@ -307,11 +307,17 @@ class LocalDatabase implements IPluginStorage<{}> {
     };
 
     const sinopiadbPath: string = dbGenPath(DEPRECATED_DB_NAME);
-    if (fs.existsSync(sinopiadbPath)) {
+    try {
+      fs.accessSync(sinopiadbPath, fs.constants.F_OK);
```
Comment (juanpicado) [inline src/local-database.ts]: Just for reference https://nodejs.org/api/fs.html#fs_file_access_constants

File: src/local-fs.ts [hunk 4]
```diff
@@ -195,7 +195,8 @@ export default class LocalFS implements ILocalFSPackageManager {
 
     const pathName: string = this._getStorage(name);
 
-    fs.exists(pathName, exists => {
+    fs.access(pathName, (fileNotFound) => {
+      const exists = !fileNotFound;
```
Comment (juanpicado) [inline src/local-fs.ts]: The approach displayed in the docs seems much more easy to understand, what do you think?
https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback
```
// Check if the file exists in the current directory.
fs.access(file, fs.constants.F_OK, (err) => {
  console.log(`${file} ${err ? 'does not exist' : 'exists'}`);
});
```

Comment (juanmanual) [reply src/local-fs.ts]: This implementation is very similar to the linked one. The fs.constants.F_OK is redundant in the original example, since it is the default mode, so I left it out. I gave the error a name because I thought it made the logic easier to follow. If you know that the error case is file-not-found, it is clear that the opposite of that is that the file exists. Since I didn't want to flip the if-else statement that follows, I had to negate the condition, otherwise we'd have to check against `!fileNotFound` which in my opinion is less intuitive.

File: src/local-fs.ts (hunk 1 above)
Comment (juanmanual) [reply src/local-fs.ts]: I changed the constant name back. I had renamed the constant so I could reuse it in a check for that error, because it seemed strange to have to switch between error codes. Since the constant was exported I assumed any code that relied on that constant would import it and so I hadn't considered that it might cause issues.

File: src/local-fs.ts (hunk 4 above)
Comment (juanpicado) [reply src/local-fs.ts]: Thanks @juanmanual 👍

Comment (juanpicado) [reply src/local-fs.ts]: One more question, about this

https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback

> Using fs.access() to check for the accessibility of a file before calling fs.open(), fs.readFile() or fs.writeFile() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file is not accessible.

We follow the `fs.access` with a `const file = fs.createWriteStream(temporalName);` which makes me thing whether the argument above also applies to `fs.createWriteStream` 🤔 . What do you think?

**Comment by juanpicado** [comment]
Thanks @juanmanual ! Really nice contribution.
